File: assignment1/assigment1/p1.m
clear
clc
close all

im1 = imread('/home/sajal/3-1/DIP/Assignments/assignment1/assigment1/spot1-diff1.jpg');
im2 = imread('/home/sajal/3-1/DIP/Assignments/assignment1/assigment1/spot1-diff2.jpg');
% figure,imshow(im1);
% figure,imshow(im2);

THRESH = 40;

imdiff = abs(im1 - im2);

[x y] = find(imdiff(:,:,1) > THRESH | imdiff(:,:,2) > THRESH | imdiff(:,:,3) > THRESH );

imnew = zeros(size(imdiff,1),size(imdiff,2));

for i=1:size(x,1)
    a=x(i,1);
    b=y(i,1);
    imnew(a,b)=255;
end

%figure,imshow(imnew);

visited=zeros(size(imnew));
leftpoint=zeros(1,2);
rightpoint=zeros(1,2);
toppoint=zeros(1,2);
bottompoint=zeros(1,2);
dimen=[];
recchar=zeros(1,4);
count=0;

for i=1:size(imnew,1)
    for j=1:size(imnew,2)
        if(imnew(i,j)==255 && visited(i,j)==0)
            imright=1;
            imtop=size(imnew,1);
            imleft=size(imnew,2);
            imbottom=1;
            [visited,left,right,top,bottom]=paintfill(imnew,visited,i,j,255,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
            breadth=abs(left(1)-right(1));
            height=abs(bottom(2)-top(2));
            area=breadth*height;
            if(area>40)
                count=count+1;
                recchar(1)=left(1);
                recchar(2)=top(2);
                recchar(3)=breadth;
                recchar(4)=height;
                if(count==1)
                    dimen=recchar;
                else
                    dimen=[dimen;recchar];
                end
                
%                 z = rectangle('Position',[left(1),top(2),breadth,height]);
%                 set(z,'EdgeColor','red');
            end
        end                                            
    end
end

figure,imshow(im1);
for i=1:size(dimen,1)
    z=rectangle('Position',[dimen(i,1),dimen(i,2),dimen(i,3),dimen(i,4)]);
    set(z,'EdgeColor',[1 0 0]);
    set(z,'LineWidth',1);
end
% 
figure,imshow(im2);
for i=1:size(dimen,1)
    z=rectangle('Position',[dimen(i,1),dimen(i,2),dimen(i,3),dimen(i,4)]);
    set(z,'EdgeColor', [1 0 0]);
    set(z,'LineWidth',1);
end

File: assignment1/assigment1/p1.py
import numpy as np
import cv2

im1 = cv2.imread('/home/sajal/Desktop/DIP/Assignments/assignment1/assigment1/spot1-diff1.jpg')
im2 = cv2.imread('/home/sajal/Desktop/DIP/Assignments/assignment1/assigment1/spot1-diff2.jpg')

def paintfill( im,vis,x,y,val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint) :
	if (x >= 1 and y >= 1 and x <= size(im,1) and y <= size(im,2) and vis(x,y) == 0 && im(x,y) == val) :
		vis(x,y)=1;

		if(y < imleft)
    		leftpoint=[y,x];
    		imleft=leftpoint(1);

		if(y > imright)
    		rightpoint=[y,x];
    		imright=rightpoint(1);

		if(x < imtop)
    		toppoint=[y,x];
    		imtop=toppoint(2);

		if(x>imbottom)
    		bottompoint=[y,x];
    		imbottom=bottompoint(2);
    
		#if(im(x,y)==val)
    	#visited(x,y)=val;
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,(x-1),y,val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,(x+1),y,val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,x,(y+1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,x,(y-1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,(x-1),(y-1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,(x+1),(y-1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,x+1,(y+1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
    	[vis,leftpoint,rightpoint,toppoint,bottompoint] = paintfill(im,vis,x-1,(y+1),val,imleft,imright,imtop,imbottom,leftpoint,rightpoint,bottompoint,toppoint);
	#else
		#varargout=1;
		#[visited]=paintfill(im,x,y,val);
		#visited(i-1,j-1)=1;
		#visited(i,j-1)=1;
		#visited(i+1,j-1)=1;
		#visited(i-1,j)=1;
		#visited(i+1,j)=1;
		#visited(i-1,j+1)=1;
		#visited(i,j+1)=1;
		#visited(i+1,j+1)=1;


#cv2.imshow('image1',im1)
#cv2.imshow('image2',im2)
#cv2.waitKey(0)
#cv2.destroyAllWindows()
THRESH = 40

diffIm = cv2.absdiff(im1, im2)
[diffImCh1, diffImCh2, diffImCh3] = cv2.split(diffIm)
height, width, channels = diffIm.shape
zerosImage = np.zeros((height,width,1), np.uint8)

ret,threshImageChnl1 = cv2.threshold(diffImCh1,THRESH,255,cv2.THRESH_BINARY)
ret,threshImageChnl2 = cv2.threshold(diffImCh2,THRESH,255,cv2.THRESH_BINARY)
ret,threshImageChnl3 = cv2.threshold(diffImCh3,THRESH,255,cv2.THRESH_BINARY)

zerosImage = threshImageChnl1 + threshImageChnl2 + threshImageChnl3

visited = np.zeros((height, width,1), np.uint8)
leftpoint = np.zeros(1,2)
rightpoint = np.zeros(1,2)
toppoint = zeros(1,2)
bottompoint = zeros(1,2)
dimen = []
recchar = zeros(1,4)
count = 0

for i in range(width) :
	for j in range(height) :

